agents/builtin/mongodb/internal/test/profiling/profiling.go

```diff
@@ -0,0 +1,185 @@
+// pmm-agent
+// Copyright (C) 2018 Percona LLC
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+package profiling
+
+import (
+	"fmt"
+
+	"github.com/percona/pmgo"
+	"gopkg.in/mgo.v2"
+	"gopkg.in/mgo.v2/bson"
+)
+
+type run func(session pmgo.SessionManager) error
+type runDB func(session pmgo.SessionManager, dbname string) error
+
+type Profiling struct {
+	url     string
+	session pmgo.SessionManager
+	err     error
+}
+
+func New(url string) *Profiling {
+	p := &Profiling{
+		url: url,
+	}
+	p.session, p.err = createSession(url)
+	return p
+}
+
+func (p *Profiling) Enable(dbname string) error {
+	return p.Run(func(session pmgo.SessionManager) error {
+		return profile(session.DB(dbname), 2)
+	})
+}
+
+func (p *Profiling) Disable(dbname string) error {
+	return p.Run(func(session pmgo.SessionManager) error {
+		return profile(session.DB(dbname), 0)
+	})
+}
+
+func (p *Profiling) Drop(dbname string) error {
+	return p.Run(func(session pmgo.SessionManager) error {
+		if !p.Exist(dbname) {
+			return nil
+		}
+		return session.DB(dbname).C("system.profile").DropCollection()
+	})
+}
+
+func (p *Profiling) Exist(dbnameToLook string) bool {
+	found := fmt.Errorf("found db: %s", dbnameToLook)
+	err := p.RunDB(func(session pmgo.SessionManager, dbname string) error {
+		if dbnameToLook == dbname {
+			return found
+		}
+		return nil
+	})
+
+	return err == found
+}
+
+func (p *Profiling) Reset(dbname string) error {
+	err := p.Disable(dbname)
+	if err != nil {
+		return err
+	}
+	err = p.Drop(dbname)
+	if err != nil {
+		return err
+	}
+	err = p.Enable(dbname)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (p *Profiling) EnableAll() error {
+	return p.RunDB(func(session pmgo.SessionManager, dbname string) error {
+		return p.Enable(dbname)
+	})
+}
+
+func (p *Profiling) DisableAll() error {
+	return p.RunDB(func(session pmgo.SessionManager, dbname string) error {
+		return p.Disable(dbname)
+	})
+}
+
+func (p *Profiling) DropAll() error {
+	return p.RunDB(func(session pmgo.SessionManager, dbname string) error {
+		return p.Drop(dbname)
+	})
+}
+
+func (p *Profiling) ResetAll() error {
```

Comment:
Let's unexport or remove all methods we don't use (all expect Start and Stop?).

Reply:
There are no `Start/Stop` methods. It's Profiler helper to enable and/or disable MongoDB profiler in tests.

Reply:
Is it used anywhere?

Reply:
I don't understand why should we remove public APIs if we don't use them. It doesn't make any sense for me. Let fork all third-party libraries we use and remove all methods we don't use from it, why not? 

Why should I drop it? Tomorrow If I would like to add some tests I need to add it one more time? No thanks. I won't do it.

Reply:
Then un-export them if you think they will be useful later.

Reply:
Why? Let's fork all libs we are using and un-export all public methods we don't use. What is the problem with that?

Reply:
There is a problem in your argument – reductio ad absurdum. The code in this repository (outside of `vendor/`) is not a third-party code. This is also not a library package. It should not expose more than necessary. Those methods are not used. I would argue we should remove them, but if you think they could be useful later – fine, let's keep them, but not export them.

And I don't think this discussion worth the time we spend on it. Please un-export those methods.

Reply:
The problem is, your arguments it's only your wishes they have no sense or value in the context of this task or review. The reason you ask about it is only that you can, not because this code doesn't work or doing something wrong. 

Also, I think it worth the time as at this point we have only one code review criteria: "You should like this changes". 

I disagree with that code review approach so I will challenge it. 

I don't understand why you get to the bottom of every unnecessary detail when we have more big problems in design and architecture. For the real problems, you always say `We can fix it later...`. But for this stupid thing you always rejecting the review.

```
This is also not a library package. It should not expose more than necessary.
```
All packages expose some public API and the caller depends on a package, not vice versa.
Which will be public is the decision of package creator, not package user.

```
fine, let's keep them, but not export them.
```
It also doesn't make any sense for me to do some methods private only because the caller doesn't use it. Where is it came from? I never have heard about that approach. It's insane.

Reply:
Also as long as we have something in a repository like this: https://github.com/percona/pmm-managed/blob/PMM-2.0/services/mysql/mysql.go IMHO you have no rights to speak about something we should remove or unexport.